docs/lazy-adr/adr-004-core-types.md

```diff
@@ -0,0 +1,216 @@
+If there are any additional fields that conflict with above's claimed minimalism, then they are necessarily inherited 
+by the ABCI imposed separation between application state machine and consensus/networking (often also referred to as ABCI-server and -client).
+Where such tradeoffs are made, we explicitly comment on them.
+
+## Detailed Design
+
+### Transactions
+
+In Optimint, like in Tendermint, Transactions are just an opaque slice of bytes:
+
+```go
+type Tx []byte
+type Txs []Tx
+```
+
+If necessary `Tx` could be turned into a struct. Currently, there is no need for that though.
+
+### Block Header
+
+```go
+type Header struct {
+    // Block and App version
+    Version Version 
+    // NamespaceID identifies this chain e.g. when connected to other rollups via IBC.
+    NamespaceID [8]byte  
+    
+    Height  uint64               
+    Time    uint64 // time in tai64 format           
+    
+    // prev block info
+    LastHeaderHash [32]byte 
+    
+    // hashes of block data
+    LastCommitHash     [32]byte // commit from aggregator(s) from the last block
+    DataHash           [32]byte // Block.Data root aka Transactions
+    ConsensusHash      [32]byte // consensus params for current block
+    AppHash            [32]byte  // state after applying txs from the current block
+    
+    // root hash of all results from the txs from the previous block
+    // This is ABCI specific but smart-contract chains require some way of committing to transaction receipts/results.
+    LastResultsHash [32]byte
+    
+    
+    // Note that the address can be derived from the pubkey which can be derived 
+    // from the signature when using secp256k.
+    // We keep this in case users choose another signature format where the 
+    // pubkey can't be recovered by the signature (e.g. ed25519).
+    ProposerAddress Address  // original proposer of the block
+}
+
+// Version captures the consensus rules for processing a block in the blockchain,
+// including all blockchain data structures and the rules of the application's
+// state transition machine.
+// This is equivalent to the tmversion.Consensus type in Tendermint.
+type Version struct { 
+    Block uint32 
+    App   uint32
+}
+```
+
+### Block and Block.Data
+
+```go
+type Block struct {
+    Header      Header
+    Data        Data
+    LastCommit  *Commit
+}
+
+type Data struct {
+    Txs Txs
+    IntermediateStateRoots IntermediateStateRoots
+    Evidence EvidenceData
+}
+
+type EvidenceData struct {
+    Evidence []Evidence
+}
+```
+
+#### Evidence
+
+`Evidence` represents a go-interface (or oneof in protobuf) of known set of concrete fraud-proofs.
+The details for this will be defined in a separated adr/PR.
+
+Here is an incomplete list of potenial evidence types:
+- Same Aggregator signed two different blocks at the same height
+  - figure out if this is actually malicious / slashable behaviour - e.g. clients could simply accept the last block included in a LL block
```

Comment:
@adlerjohn should we consider this misbehavior or not? I'm not sure.

Reply:
For now, I guess so. I'll have to think about it more, but it might be dependent on the particular choice of leader selection.